src/bgw/job.c

```diff
@@ -176,84 +180,168 @@ ts_bgw_job_get_all(size_t alloc_size, MemoryContext mctx)
 static void
-bgw_job_scan_one(int indexid, ScanKeyData scankey[], int nkeys, tuple_found_func tuple_found,
-				 tuple_filter_func tuple_filter, void *data, MemoryContext mctx, LOCKMODE lockmode,
-				 bool fail_if_not_found)
+init_scan_by_job_id(ScanIterator *iterator, int32 job_id)
 {
-	Catalog *catalog = ts_catalog_get();
-	ScannerCtx scanctx = {
-		.table = catalog_get_table_id(catalog, BGW_JOB),
-		.index = catalog_get_index(catalog, BGW_JOB, indexid),
-		.nkeys = nkeys,
-		.scankey = scankey,
-		.tuple_found = tuple_found,
-		.filter = tuple_filter,
-		.data = data,
-		.lockmode = lockmode,
-		.scandirection = ForwardScanDirection,
-		.result_mctx = mctx,
-	};
-
-	ts_scanner_scan_one(&scanctx, fail_if_not_found, "bgw job");
+	iterator->ctx.index = catalog_get_index(ts_catalog_get(), BGW_JOB, BGW_JOB_PKEY_IDX);
+	ts_scan_iterator_scan_key_init(iterator,
+								   Anum_bgw_job_pkey_idx_id,
+								   BTEqualStrategyNumber,
+								   F_INT4EQ,
+								   Int32GetDatum(job_id));
 }
 
-static inline void
-bgw_job_scan_job_id(int32 bgw_job_id, tuple_found_func tuple_found, tuple_filter_func tuple_filter,
-					void *data, MemoryContext mctx, LOCKMODE lockmode, bool fail_if_not_found)
-{
-	ScanKeyData scankey[1];
+/* Lock a job tuple using a heavyweight tuple lock. Heavyweight tuple locks are
+ *  used to lock the job row while a job is running to prevent a job from being
+ *  modified while in the middle of a run. Note that in regular DML operations,
+ *  heavyweight locks on tuples are only sometimes taken as otherwise the MVCC
+ *  system uses tuple headers for its locking. Thus, we need to take this lock
+ *  explicitly ourselves during DML and treat this more as an advisory lock system
+ *  managed by ourselves.
+ *
+ * returns whether or not the lock was obtained (false return only possible if dontwait==false)
+ */
 
-	ScanKeyInit(&scankey[0],
-				Anum_bgw_job_stat_pkey_idx_job_id,
-				BTEqualStrategyNumber,
-				F_INT4EQ,
-				Int32GetDatum(bgw_job_id));
-	bgw_job_scan_one(BGW_JOB_STAT_PKEY_IDX,
-					 scankey,
-					 1,
-					 tuple_found,
-					 tuple_filter,
-					 data,
-					 mctx,
-					 lockmode,
-					 fail_if_not_found);
+static bool
+lock_tuple(HeapTuple heap, Relation relation, LOCKMODE mode, bool session_lock, LOCKTAG *tag,
+		   bool block)
+{
+	ItemPointer tid = &heap->t_self;
+	SET_LOCKTAG_TUPLE(*tag,
+					  relation->rd_lockInfo.lockRelId.dbId,
+					  relation->rd_lockInfo.lockRelId.relId,
+					  ItemPointerGetBlockNumber(tid),
+					  ItemPointerGetOffsetNumber(tid));
+
+	return LockAcquire(tag, mode, session_lock, !block) != LOCKACQUIRE_NOT_AVAIL;
 }
 
-static ScanTupleResult
-bgw_job_tuple_found(TupleInfo *ti, void *const data)
+static BgwJob *
+ts_bgw_job_find_with_lock(int32 bgw_job_id, MemoryContext mctx, LOCKMODE tuple_lock_mode,
+						  LOCKTAG *session_lock_locktag, bool block)
 {
-	BgwJob **job_pp = data;
+	/* Take a share lock on the table to prevent concurrent data changes during scan. This lock will
```

Comment:
No, this share lock only prevents data during the scan. It will be released after the scan in either case. `lock_job` actually does the session lock stuff